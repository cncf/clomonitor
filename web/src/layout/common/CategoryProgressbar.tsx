import { isUndefined } from 'lodash';
import { useLocation, useNavigate } from 'react-router-dom';

import getCategoryColor from '../../utils/getCategoryColor';
import roundScoreValue from '../../utils/roundScoreValue';
import styles from './CategoryProgressbar.module.css';

interface Props {
  icon?: JSX.Element;
  value?: number;
  name: string;
  bigSize?: boolean;
  linkTo?: string;
  scrollIntoView?: (id?: string) => void;
}

const CategoryProgressbar = (props: Props) => {
  const navigate = useNavigate();
  const location = useLocation();
  const color = getCategoryColor(props.value);

  return (
    <div className={`${styles.wrapper} ${props.bigSize ? 'col-12 col-lg-9 col-xxxl-8' : 'col-12'}`}>
      <div className="d-flex flex-row bg-white position-relative border overflow-hidden">
        <div
          className={`d-flex flex-row align-items-center text-muted fw-bold flex-nowrap px-1 my-auto ${styles.title}`}
        >
          {props.icon && <span className={`pe-1 d-inline-block position-relative ${styles.icon}`}>{props.icon}</span>}
          {!isUndefined(props.linkTo) ? (
            <button
              className={`btn btn-link text-truncate text-muted fw-bold p-0 text-decoration-none ${styles.btn}`}
              onClick={() => {
                if (props.scrollIntoView) {
                  props.scrollIntoView(`#${props.linkTo}`);
                }
                navigate(
                  {
                    pathname: location.pathname,
                    hash: props.linkTo,
                  },
                  { state: location.state }
                );
              }}
              aria-label={`Go from summary to section: ${props.linkTo}`}
            >
              {props.name}
            </button>
          ) : (
            <span className="text-truncate">{props.name}</span>
          )}
        </div>
        <div className={`text-center fw-bold font-monospace ${styles.value} ${props.bigSize ? styles.bigSize : ''}`}>
          {isUndefined(props.value) ? 'n/a' : roundScoreValue(props.value)}
        </div>
        <div
          className={`flex-grow-1 position-relative mx-2 ${styles.progressWrapper}  ${
            props.bigSize ? styles.progressBigWrapper : ''
          }`}
        >
          {!isUndefined(props.value) && (
            <div
              data-testid="line"
              className={`position-absolute start-0 top-0 bottom-0 ${styles.line}`}
              style={{
                width: `${props.value || 1}%`,
                backgroundColor: `var(--rm-${color})`,
              }}
            />
          )}
        </div>
      </div>
    </div>
  );
};

export default CategoryProgressbar;
