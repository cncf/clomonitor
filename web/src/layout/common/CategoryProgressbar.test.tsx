import { render, screen } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';

import CategoryProgressbar from './CategoryProgressbar';

const defaultProps = {
  value: 80,
  name: 'Documentation',
};

describe('CategoryProgressbar', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  it('creates snapshot', () => {
    const { asFragment } = render(
      <Router>
        <CategoryProgressbar {...defaultProps} />
      </Router>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders proper content', () => {
    render(
      <Router>
        <CategoryProgressbar {...defaultProps} />
      </Router>
    );
    expect(screen.getByText('Documentation')).toBeInTheDocument();
    expect(screen.getByText('80')).toBeInTheDocument();

    const line = screen.getByTestId('line');
    expect(line).toBeInTheDocument();
    expect(line).toHaveStyle('width: 80%');
  });
});
