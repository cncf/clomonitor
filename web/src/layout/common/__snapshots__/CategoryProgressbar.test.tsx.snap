// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoryProgressbar creates snapshot 1`] = `
<DocumentFragment>
  <div
    class="wrapper col-12"
  >
    <div
      class="d-flex flex-row bg-white position-relative border border-1 overflow-hidden"
    >
      <div
        class="d-flex flex-row align-items-center text-muted fw-bold flex-nowrap px-1 my-auto title"
      >
        <span
          class="text-truncate"
        >
          Documentation
        </span>
      </div>
      <div
        class="text-center fw-bold font-monospace value "
      >
        80
      </div>
      <div
        class="flex-grow-1 position-relative mx-2 progressWrapper  "
      >
        <div
          class="position-absolute start-0 top-0 bottom-0 line"
          data-testid="line"
          style="width: 80%;"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
